File: sdk/javascript/src/types/vc-registry.ts
/**
 * Verifiable Credential status enum
 */
export enum VCStatus {
  ACTIVE = 0,
  REVOKED = 1,
  EXPIRED = 2,
  SUSPENDED = 3,
}

/**
 * Verifiable Credential
 */
export interface VerifiableCredential {
  id: string;
  issuer: string;
  subject: string;
  type: string[];
  issuanceDate: Date;
  expirationDate?: Date;
  claims: Record<string, any>;
  proof: {
    type: string;
    created: Date;
    proofPurpose: string;
    verificationMethod: string;
    signature: string;
  };
  status: VCStatus;
  revocationReason?: string;
  revokedAt?: Date;
}

/**
 * Mint VC parameters
 */
export interface MintVCParams {
  issuer: string;
  subject: string;
  credentialType: string[];
  claims: Record<string, any>;
  expirationDate?: Date;
  transferable?: boolean;
}

/**
 * Revoke VC parameters
 */
export interface RevokeVCParams {
  vcId: string;
  issuer: string;
  reason: string;
}

/**
 * Verifiable Presentation
 */
export interface VerifiablePresentation {
  id: string;
  holder: string;
  verifiableCredentials: string[]; // VC IDs
  type: string[];
  purpose: string;
  createdAt: Date;
  expiresAt?: Date;
  proof: {
    type: string;
    created: Date;
    proofPurpose: string;
    verificationMethod: string;
    signature: string;
    challenge?: string;
  };
}

/**
 * Create presentation parameters
 */
export interface CreatePresentationParams {
  holder: string;
  vcIds: string[];
  purpose: string;
  expirationDuration?: number;
  challenge?: string;
}

/**
 * Verification result
 */
export interface VCVerificationResult {
  valid: boolean;
  vcId: string;
  issuer: string;
  subject: string;
  status: VCStatus;
  verified: boolean;
  signatureValid: boolean;
  notExpired: boolean;
  notRevoked: boolean;
  issuerTrusted: boolean;
  errors: string[];
  verifiedAt: Date;
}

/**
 * Presentation verification result
 */
export interface PresentationVerificationResult {
  valid: boolean;
  presentationId: string;
  holder: string;
  vcResults: VCVerificationResult[];
  signatureValid: boolean;
  notExpired: boolean;
  challengeVerified: boolean;
  errors: string[];
  verifiedAt: Date;
}

/**
 * VC Registry statistics
 */
export interface VCStats {
  totalVCs: number;
  activeVCs: number;
  revokedVCs: number;
  expiredVCs: number;
  totalIssuers: number;
  totalSubjects: number;
  totalPresentations: number;
  credentialTypes: {
    type: string;
    count: number;
  }[];
}

/**
 * VC query filters
 */
export interface VCQueryFilters {
  issuer?: string;
  subject?: string;
  type?: string;
  status?: VCStatus;
  fromDate?: Date;
  toDate?: Date;
  limit?: number;
  offset?: number;
}

/**
 * VC Registry parameters
 */
export interface VCRegistryParams {
  issuanceEnabled: boolean;
  revocationEnabled: boolean;
  maxExpirationDuration: number;
  requireExpirationDate: boolean;
  trustedIssuers: string[];
  supportedTypes: string[];
  verificationFee: string;
}

/**
 * Issuer info
 */
export interface IssuerInfo {
  address: string;
  name: string;
  credentialsIssued: number;
  credentialsRevoked: number;
  trusted: boolean;
  registeredAt: Date;
  publicKey: string;
}

File: sdk/python/aura/types/vc_registry.py
"""Type definitions for VCRegistry (Verifiable Credentials) module."""

from dataclasses import dataclass
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class VCStatus(Enum):
    """Verifiable Credential status."""

    ACTIVE = "active"
    REVOKED = "revoked"
    EXPIRED = "expired"
    SUSPENDED = "suspended"


class VCType(Enum):
    """Types of verifiable credentials."""

    IDENTITY = "identity"
    EDUCATION = "education"
    EMPLOYMENT = "employment"
    LICENSE = "license"
    CERTIFICATION = "certification"
    CUSTOM = "custom"


class PresentationStatus(Enum):
    """Presentation verification status."""

    VERIFIED = "verified"
    INVALID = "invalid"
    EXPIRED = "expired"
    REVOKED = "revoked"


@dataclass
class VCParams:
    """Parameters for creating a verifiable credential."""

    issuer: str
    subject: str
    vc_type: VCType
    claims: Dict[str, Any]
    expiry_date: Optional[datetime] = None
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class VerifiableCredential:
    """Verifiable Credential information."""

    vc_id: str
    issuer: str
    subject: str
    vc_type: VCType
    status: VCStatus
    claims: Dict[str, Any]
    proof: str
    issued_at: datetime
    expires_at: Optional[datetime] = None
    revoked_at: Optional[datetime] = None
    metadata: Optional[Dict[str, Any]] = None
    schema_id: Optional[str] = None


@dataclass
class VCProof:
    """Credential cryptographic proof."""

    proof_type: str
    proof_value: str
    verification_method: str
    created: datetime
    proof_purpose: str
    challenge: Optional[str] = None
    domain: Optional[str] = None


@dataclass
class VCPresentation:
    """Verifiable Presentation."""

    presentation_id: str
    holder: str
    credentials: List[str]
    proof: VCProof
    created_at: datetime
    expires_at: Optional[datetime] = None
    challenge: Optional[str] = None
    domain: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class PresentationParams:
    """Parameters for creating a presentation."""

    holder: str
    credential_ids: List[str]
    challenge: Optional[str] = None
    domain: Optional[str] = None
    purpose: str = "authentication"


@dataclass
class VCRevocationParams:
    """Parameters for revoking a credential."""

    vc_id: str
    issuer: str
    reason: str
    proof: str


@dataclass
class VCSchema:
    """Credential schema definition."""

    schema_id: str
    name: str
    version: str
    vc_type: VCType
    required_fields: List[str]
    optional_fields: List[str]
    field_types: Dict[str, str]
    created_at: datetime
    author: str


@dataclass
class IssuerInfo:
    """Credential issuer information."""

    issuer_id: str
    name: str
    did: str
    public_key: str
    verified: bool
    total_issued: int
    total_revoked: int
    reputation_score: float
    registered_at: datetime


@dataclass
class VCVerificationResult:
    """Credential verification result."""

    vc_id: str
    valid: bool
    status: VCStatus
    issuer_verified: bool
    signature_valid: bool
    not_expired: bool
    not_revoked: bool
    verified_at: datetime
    errors: List[str]
    warnings: List[str]


@dataclass
class RegistryStats:
    """Registry statistics."""

    total_credentials: int
    active_credentials: int
    revoked_credentials: int
    expired_credentials: int
    total_issuers: int
    total_holders: int
    total_presentations: int
    credentials_by_type: Dict[str, int]


@dataclass
class VCQuery:
    """Query parameters for searching credentials."""

    issuer: Optional[str] = None
    subject: Optional[str] = None
    vc_type: Optional[VCType] = None
    status: Optional[VCStatus] = None
    issued_after: Optional[datetime] = None
    issued_before: Optional[datetime] = None
    include_revoked: bool = False
